File: 09-15-iterators/adts/list/array_list.cpp
#pragma once

#include "list.cpp"
#include <assert.h>

template <class T> class arrayList : public list<T> {
private:
  int capacity;
  int count;
  T *arr;

  class arrListIterator:public iterator<T>{
  private:
    int pos;
    int count;
    T* arr;
  public:
    arrListIterator(T*arr, int count){
      this->pos = 0;
      this->count = count;
      this->arr = arr;
    }

    virtual bool hasNext() override{
      return this->pos < this->count;
    }
    virtual T next() override{
      T elem = this->arr[pos];
      pos++;
      return elem;
    }
  };

  void grow() {
    int newCapacity = this->capacity * 2;
    T *newArr = new T[newCapacity];
    for (int i = 0; i < this->capacity; i++) {
      newArr[i] = arr[i];
    }

    this->arr = newArr;
    this->capacity = newCapacity;
  }

public:
  arrayList(int expectedSize){
    this->arr = new T[expectedSize];
    this->capacity = expectedSize;
  }

  virtual void add(T data) override {
    if (this->count == this->capacity) {
      grow();
    }

    this->arr[count] = data;
    this->count++;
  }
  virtual int size() override { assert(false); }
  virtual T getPos(int p) override {
    assert(p < count);
    return this->arr[p];
  }
  virtual int findPos(T data) override { assert(false); }
  virtual void remove(T data) override { assert(false); }
  virtual void removePos(int pos) { assert(false); }

  virtual iterator<T>* getIterator(){
    return new arrListIterator(this->arr, this->count);
  }
};

File: 09-15-iterators/adts/list/ArrayList.java
package adts.list;

import java.util.Iterator;

public class ArrayList<T> implements List<T> {
    private int count;
    private Object[] arr;

    class arrListIterator implements Iterator<T> {
        private int pos;
        private int count;
        private Object[] arr;

        arrListIterator(Object[] arr, int count) {
            this.pos = 0;
            this.count = count;
            this.arr = arr;
        }

        @Override
        public boolean hasNext() {
            return this.pos < this.count;
        }

        @Override
        public T next() {
            T elem = (T) this.arr[pos];
            pos++;
            return elem;
        }

    }

    private void grow() {
        int newCapacity = this.arr.length * 2;
        Object[] newArr = new Object[newCapacity];
        for (int i = 0; i < this.arr.length; i++) {
            newArr[i] = arr[i];
        }

        this.arr = newArr;
    }

    public ArrayList(int expectedSize) {
        this.arr = new Object[expectedSize];
    }

    @Override
    public void add(T data) {
        if (this.count == this.arr.length) {
            grow();
        }

        this.arr[count] = data;
        this.count++;
    }

    @Override
    public int size() {
        throw new RuntimeException("unimplemented method");
    }

    @Override
    public T getPos(int p) {
        assert (p < count);
        return (T) this.arr[p];
    }

    @Override
    public int findPos(T data) {
        throw new RuntimeException("unimplemented method");
    }

    @Override
    public void remove(T data) {
        throw new RuntimeException("unimplemented method");
    }

    @Override
    public void removePos(int pos) {
        throw new RuntimeException("unimplemented method");
    }

    @Override
    public Iterator<T> iterator() {
        return new arrListIterator(this.arr, this.count);
    }

}
